Ahead/behind show up-to-date and gone

gerhardol opened the pull request:
First part of #8553 


## Proposed changes

- Show ahead/behind if remote is up to date
- Show if the tracking branch no longer exists "gone"

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

Only the branch with differences displays the count
![branch-before](https://user-images.githubusercontent.com/6248932/96507421-6de44d80-1259-11eb-84e5-3983b5a3fd2b.JPG)

No indication if the branches are up to date
![push-none](https://user-images.githubusercontent.com/6248932/96507433-72a90180-1259-11eb-8e7c-27a5cec1171f.JPG)

### After

feature/i8553-ahead-zero-gone is not yet pushed
feature/i8548-multiple-force-push is deleted at the remote
feature/i8448-presentation is pushed and in sync with the remote
feature/i8448-range-diff-3 same as above (not displayed)
![branch-after](https://user-images.githubusercontent.com/6248932/96507774-efd47680-1259-11eb-8296-5c56362b63b1.JPG)

After a branch is pushed
![push-after](https://user-images.githubusercontent.com/6248932/96507812-febb2900-1259-11eb-8ea2-1a725b4b4ad2.JPG)


## Test methodology 

Tests adapted and expanded

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented on GitCommands/Git/AheadBehindData.cs:
```diff
@@ -2,15 +2,21 @@
 {
     public struct AheadBehindData
     {
+        public static readonly string Gone = "gone";
```
Make translatable or replace with a character, e.g. a dash.

gerhardol replied:
(What I was most unsure about.)
"gone" is what the Git command writes, but I prefer "-" myself too.
(The change is on line 13, the const is still needed.)

gerhardol replied:
> I was going to ask what git uses, but you've already said. If "gone" is
git's lingo I'm ok with it. Though my first reaction was to have "deleted"
or "removed".
"-" to me is akin "n/a", which is probably not really suitable in here. I
think this may be appropriate for "not pushed" case....

@RussKie 
Changing to "-" right now
Should we have this translatable if sticking with "gone"?

For branches that have lost the tracking (second part of #8553), should we reserve something?
(I want to display this in FormPush at least)

RussKie replied:
If it's a git's string I think we can have it as is. Need to have a comment in the code why it is not translated though.

gerhardol replied:
for-each-ref is a plumbing command and should not be translated, otherwise also ahead/behind could be incorrect.
I added one more check (made the parsing even more complicated) to handle garbage, now it could be interpreted as "0" changes.

I checked the Git code and 'gone' as well as 'ahead'/'behind' can be translated in other printouts. I tried to get command cli to be presented in something else than en, but that failed (Just "LANGUAGE=de_DE git status" should be enough). Maybe not included with git-for-windows? So maybe it should be translated later. But the translated "expression" can be harder to understand, git gone yields some Google hits.

Not sure if the comment is good though...

mstv commented on GitCommands/Git/AheadBehindData.cs:
```diff
@@ -2,15 +2,21 @@
+                ? Gone
+                : AheadCount == "0" && string.IsNullOrEmpty(BehindCount)
+                ? "0↑↓"
+                : (!string.IsNullOrEmpty(AheadCount) && AheadCount != "0"
+                    ? AheadCount + "↑" + (!string.IsNullOrEmpty(BehindCount) ? " " : string.Empty)
+                    : string.Empty)
+                    + (!string.IsNullOrEmpty(BehindCount) ? BehindCount + "↓" : string.Empty);
```
Replace with a local function using StringBuilder or
```suggestion
                : (!string.IsNullOrEmpty(AheadCount) && AheadCount != "0"
                   ? AheadCount + "↑" + (!string.IsNullOrEmpty(BehindCount) ? " " : string.Empty)
                   : string.Empty)
                  + (!string.IsNullOrEmpty(BehindCount) ? BehindCount + "↓" : string.Empty);
```

gerhardol replied:
Changing the indentation

mstv commented on GitCommands/Git/AheadBehindDataProvider.cs:
```diff
@@ -64,22 +69,39 @@ public AheadBehindDataProvider(Func<IExecutable> getGitExecutable)
-                        AheadCount = match.Groups["ahead_p"].Success ? match.Groups["ahead_p"].Value : match.Groups["ahead_u"].Value,
+                        // The information is displayed in the push button, so the push info is preferred (may differ from upstream)
+                        Branch = branch,
+                        AheadCount =
+
+                            // Prefer push to upstream for the count
+                            match.Groups["ahead_p"].Success
+                            ? match.Groups["ahead_p"].Value
+                            : match.Groups["ahead_u"].Success
+                            ? match.Groups["ahead_u"].Value
+
+                            // No information about the remote branch, it is gone
+                            : match.Groups["gone_p"].Success || match.Groups["gone_u"].Success
+                            ? AheadBehindData.Gone
+
+                            // A remote exists, but "track" does not display the count if ahead/behind match
+                            : "0",
+
```
```suggestion
                        AheadCount =
                            // Prefer push to upstream for the count
                            match.Groups["ahead_p"].Success
                            ? match.Groups["ahead_p"].Value
                            : match.Groups["ahead_u"].Success
                              ? match.Groups["ahead_u"].Value
                              : // No information about the remote branch, it is gone
                                match.Groups["gone_p"].Success || match.Groups["gone_u"].Success
                                ? AheadBehindData.Gone
                                : // A remote exists, but "track" does not display the count if ahead/behind match
                                  "0",
```

gerhardol replied:
There must be a newline before a comment....

RussKie replied:
Right now it is not very readable... You can suppress the warning using #pragma

mstv replied:
> There must be a newline before a comment....

Or use `////`.

gerhardol replied:
I used #pragma instead of misusing ////
Even more conditions now, I prefer to not right shift (should be 4 spaces if done)

mstv commented on GitCommands/Git/AheadBehindDataProvider.cs:
```diff
@@ -64,22 +69,39 @@ public AheadBehindDataProvider(Func<IExecutable> getGitExecutable)
             foreach (Match match in matches)
             {
+                var branch = match.Groups["branch"].Value;
+                var remoteRef = (match.Groups["remote_p"].Success && !string.IsNullOrEmpty(match.Groups["remote_p"].Value))
+                            ? match.Groups["remote_p"].Value
+                            : match.Groups["remote_u"].Value;
```
```suggestion
                var remoteRef = (match.Groups["remote_p"].Success && !string.IsNullOrEmpty(match.Groups["remote_p"].Value))
                                ? match.Groups["remote_p"].Value
                                : match.Groups["remote_u"].Value;
```

gerhardol replied:
why one more level?

mstv replied:
I'd either `?` align it with the start of the condition or else use one indent level.

gerhardol replied:
One indent from 'var'

mstv reviewed (commented):
Looks sensible, haven't checked in detail. the logic is checked by tests.

RussKie commented:
I was going to ask what git uses, but you've already said. If "gone" is
git's lingo I'm ok with it. Though my first reaction was to have "deleted"
or "removed".
"-" to me is akin "n/a", which is probably not really suitable in here. I
think this may be appropriate for "not pushed" case....

On Wed, 21 Oct. 2020, 9:58 am Gerhard Olsson, <notifications@github.com>
wrote:

> *@gerhardol* commented on this pull request.
> ------------------------------
>
> In GitCommands/Git/AheadBehindData.cs
> <https://github.com/gitextensions/gitextensions/pull/8562#discussion_r508888628>
> :
>
> > @@ -2,15 +2,21 @@
>  {
>      public struct AheadBehindData
>      {
> +        public static readonly string Gone = "gone";
>
> (What I was most unsure about.)
> "gone" is what the Git command writes, but I prefer "-" myself too.
> (The change is on line 13, the const is still needed.)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8562#discussion_r508888628>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXRVJOHLO5HHSCWG2ETSLYIZPANCNFSM4SY3X57Q>
> .
>

gerhardol commented:
The parsing/interpretation of the Git output is a little too complicated right now
Adding properties for gone and zero changes will simplify, or maybe set with fewer conditions and then correct after (if zero changes, overwrite afterwards).

gerhardol commented:
> The parsing/interpretation of the Git output is a little too complicated right now

Rewriting will not really improve, I suggest it is kept as is with comments

Squashed and rebased without changes
